# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
import pandas as pd
url ="https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

company_df = pd.read_csv(url)

In [3]:
company_df

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
4003,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4004,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4006,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [6]:
company_df.shape

(4008, 11)

In [8]:
company_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   str    
 1   ST                         1071 non-null   str    
 2   GENDER                     954 non-null    str    
 3   Education                  1071 non-null   str    
 4   Customer Lifetime Value    1068 non-null   str    
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   str    
 8   Policy Type                1071 non-null   str    
 9   Vehicle Class              1071 non-null   str    
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), str(8)
memory usage: 417.4 KB


Las columnas que necesitan cambio son Customer Lifetime Value y Number of Open Complaints:
- CLV: Cambiar a float. 
- Number of Open Complaints: cambiar a Int. 

In [9]:
company_df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [ ]:
company_df["ST"].unique()

<ArrowStringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

ST mezcla nombres completos y abreviaturas. Se propone unificarlos usando los nombres completos. Tras esa normalización, quedarían cinco estados: Washington, Arizona, Nevada, California y Oregon. También hay valores ausentes.

In [ ]:
company_df["GENDER"].unique()

<ArrowStringArray>
[nan, 'F', 'M', 'Femal', 'Male', 'female']
Length: 6, dtype: str

Presenta inconsistencias de escritura y valores ausentes. Se propone unificar las variantes equivalentes en "F" y "M", manteniendo los ausentes como tales.

In [ ]:
company_df["Education"].unique()

<ArrowStringArray>
[              'Master',             'Bachelor', 'High School or Below',
              'College',            'Bachelors',               'Doctor',
                    nan]
Length: 7, dtype: str

Contiene seis valores distintos, pero "Bachelor" y "Bachelors" parecen ser variantes de una misma categoría. Se propone unificarlas como "Bachelor", dejando cinco niveles educativos. También hay valores ausentes (nan)

In [ ]:
company_df["Policy Type"].unique()

<ArrowStringArray>
['Personal Auto', 'Corporate Auto', 'Special Auto', nan]
Length: 4, dtype: str

Contiene tres tipos de póliza: "Personal Auto", "Corporate Auto" y "Special Auto". No se observan categorías duplicadas por diferencias de escritura. También hay valores ausentes (nan).

Vehicle Class contiene seis categorías: coches de cuatro puertas, coches de dos puertas, SUV, SUV de lujo, deportivos y coches de lujo. También presenta valores ausentes (nan).

In [21]:
print(company_df.select_dtypes(include='number').max())
print(company_df.select_dtypes(include='number').min())

Income                  99960.000000
Monthly Premium Auto    35354.000000
Total Claim Amount       2893.239678
dtype: float64
Income                   0.000000
Monthly Premium Auto    61.000000
Total Claim Amount       0.382107
dtype: float64


In [22]:
company_df["Number of Open Complaints"].unique()

<ArrowStringArray>
['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan]
Length: 7, dtype: str

Los ingresos presentan valores entre 0 y 99 960. La prima mensual máxima de 35 354 merece revisión por su magnitud, aunque este dato por sí solo no demuestra que sea un error. El importe total de siniestros oscila entre aproximadamente 0,38 y 2 893,24.

In [28]:
company_df.select_dtypes(include="number").describe()


,Income,Monthly Premium Auto,Total Claim Amount
count,1071.000000,1071.000000,1071.000000
mean,39295.701214,193.234360,404.986909
std,30469.427060,1601.190369,293.027260
min,0.000000,61.000000,0.382107
25%,14072.000000,68.000000,202.157702
50%,36234.000000,83.000000,354.729129
75%,64631.000000,109.500000,532.800000
max,99960.000000,35354.000000,2893.239678


In [29]:
company_df.select_dtypes(include="number").mode()

,Income,Monthly Premium Auto,Total Claim Amount
0,0.0,65.0,321.6


Income — Ingresos

La media es 39.295,70, la mediana 36.234 y la moda 0. Los ingresos presentan una dispersión considerable, con una desviación estándar de 30.469,43. El 50 % central se encuentra entre 14.072 y 64.631, y el rango va de 0 a 99.960. La media superior a la mediana sugiere influencia de ingresos altos. Como 0 es el valor más frecuente, conviene comprobar si representa ausencia de ingresos o información no registrada.

Monthly Premium Auto — Prima mensual

La media es 193,23, la mediana 83 y la moda 65. El 50 % central de las primas está entre 68 y 109,50, mientras que el máximo alcanza 35.354. La gran diferencia entre media y mediana, junto con una desviación estándar de 1.601,19, indica una fuerte influencia de valores extremos altos. Estos deben revisarse; la mediana resulta más representativa del importe central que la media.

Total Claim Amount — Importe total de siniestros

La media es 404,99, la mediana 354,73 y la moda 321,60. La desviación estándar es 293,03, y el 50 % central se sitúa entre 202,16 y 532,80. Los importes oscilan entre 0,38 y 2.893,24. La media superior a la mediana y el máximo elevado sugieren una distribución con una cola hacia los importes altos.

Conclusión general

Las tres columnas tienen 1.071 valores no ausentes. La prima mensual presenta los valores extremos más destacados respecto a su distribución central. Estas conclusiones cubren las columnas numéricas analizadas; CLV y Number of Open Complaints quedan pendientes de interpretar y convertir a formato numérico.

In [30]:
categorical_columns = [
    "ST", "GENDER", "Education", "Policy Type", "Vehicle Class"
]

In [31]:
company_df[categorical_columns].describe()

,ST,GENDER,Education,Policy Type,Vehicle Class
count,1071,954,1071,1071,1071
unique,8,5,6,3,6
top,Oregon,F,Bachelor,Personal Auto,Four-Door Car
freq,320,457,324,780,576


- ST: presenta 8 valores distintos. "Oregon" es el más frecuente, con 320 registros. Sin embargo, hay variantes como "Cali" y "California", por lo que conviene unificarlas antes de comparar las frecuencias definitivas por estado.

- GENDER: tiene 954 valores no ausentes y 5 valores distintos. "F" es el más frecuente, con 457 registros, pero las variantes "Femal" y "female" se cuentan por separado. Por eso, esa frecuencia no representa el total de su categoría una vez normalizada. Tiene 117 datos informados menos que las otras columnas.

- Education: presenta 6 valores distintos. "Bachelor" es el más frecuente, con 324 registros. Unificar "Bachelor" y "Bachelors" reduciría las categorías a cinco y aumentaría la frecuencia de ese nivel educativo.

- Policy Type: contiene 3 categorías. Predomina "Personal Auto", con 780 registros, aproximadamente el 72,8 % de los valores no ausentes.

- Vehicle Class: contiene 6 categorías. "Four-Door Car" es la más frecuente, con 576 registros, aproximadamente el 53,8 % de los valores no ausentes.

Conclusión general: predominan las pólizas personales y los coches de cuatro puertas. Las columnas de estado, género y educación necesitan normalizar sus etiquetas para obtener recuentos fiables por categoría. count excluye los valores ausentes, por lo que estos resultados no permiten afirmar que las columnas con 1.071 registros estén completas sin compararlas con el total de filas.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [36]:
location_counts = company_df["ST"].value_counts()
sorted_locations = location_counts.sort_values(ascending=True)
least_common_locations = sorted_locations.head(5)
print(least_common_locations)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


El resultado corresponde a los datos sin limpiar. "AZ", "WA" y "Cali" son variantes de otros nombres de estado, por lo que deben normalizarse para obtener las frecuencias reales por estado

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [37]:
policy_counts = company_df["Policy Type"].value_counts()
most_sold_policy = policy_counts.idxmax()
print(most_sold_policy)

Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [38]:
personal_auto = company_df.loc[
    company_df["Policy Type"] == "Personal Auto"
]

corporate_auto = company_df.loc[
    company_df["Policy Type"] == "Corporate Auto"
]

personal_income_mean = personal_auto["Income"].mean()
corporate_income_mean = corporate_auto["Income"].mean()
print("Ingresos medios de Personal Auto:", personal_income_mean)
print("Ingresos medios de Corporate Auto:", corporate_income_mean)

Ingresos medios de Personal Auto: 38180.69871794872
Ingresos medios de Corporate Auto: 41390.31196581197


Los clientes con pólizas Personal Auto presentan unos ingresos medios de 38.180,70, frente a 41.390,31 de los clientes con Corporate Auto. La diferencia es de aproximadamente 3.209,61 a favor de Corporate Auto. Esto describe una diferencia entre las medias de ambos grupos, no necesariamente entre todos sus clientes.

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [40]:
print("Estadísticas generales:")
print(company_df["Total Claim Amount"].describe())

Estadísticas generales:
count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64


In [41]:
percentile_75 = company_df["Total Claim Amount"].quantile(0.75)
print("\nPercentil 75:", percentile_75)


Percentil 75: 532.8


In [42]:
high_claim_customers = company_df.loc[
    company_df["Total Claim Amount"] > percentile_75
]

In [43]:
print("\nClientes con importes elevados:", len(high_claim_customers))


Clientes con importes elevados: 264


In [44]:
print("\nEstadísticas del grupo con importes elevados:")
print(high_claim_customers.describe())


Estadísticas del grupo con importes elevados:
             Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678


In [45]:
high_claim_customers.head()

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
17,OE15005,Cali,NaN,College,394524.16%,28855.0,101.0,1/0/00,Personal Auto,SUV,647.442031
23,TZ98966,Nevada,NaN,Bachelor,245019.10%,0.0,73.0,1/3/00,Corporate Auto,Four-Door Car,554.376763
26,US89481,California,NaN,Bachelor,394637.21%,0.0,111.0,1/0/00,Personal Auto,Four-Door Car,799.200000
